api/client/contextdialer.go

```diff
@@ -38,15 +42,20 @@ func (f ContextDialerFunc) DialContext(ctx context.Context, network, addr string
 	return f(ctx, network, addr)
 }
 
-// NewAddrDialer makes a new dialer from a list of addresses
-func NewAddrDialer(addrs []string, keepAliveInterval, dialTimeout time.Duration) (ContextDialer, error) {
-	if len(addrs) == 0 {
-		return nil, trace.BadParameter("no addreses to dial")
-	}
-	dialer := net.Dialer{
+// NewDialer makes a new dialer.
+func NewDialer(keepAliveInterval, dialTimeout time.Duration) ContextDialer {
+	return &net.Dialer{
 		Timeout:   dialTimeout,
 		KeepAlive: keepAliveInterval,
 	}
+}
+
+// NewAddrsDialer makes a new dialer from a list of addresses.
+func NewAddrsDialer(addrs []string, keepAliveInterval, dialTimeout time.Duration) (ContextDialer, error) {
+	if len(addrs) == 0 {
+		return nil, trace.BadParameter("no addresses to dial")
+	}
+	dialer := NewDialer(keepAliveInterval, dialTimeout)
 	return ContextDialerFunc(func(ctx context.Context, network, _ string) (conn net.Conn, err error) {
 		for _, addr := range addrs {
```

Comment:
feels like this `addrs` fan-out dialing should be parallelized at a higher level, same as other connection attempts
`NewAddrsDialer` is only used in `connectInBackground`, which always calls it with `constants.APIDomain` as address and that address gets ignored here in favor of `addrs`.

Maybe `connectInBackground` should iterate over `cfg.Addrs` itself and `NewAddrsDialer` can be removed?

Reply:
`NewAddrDialer` is necessary for the HTTP client. It looks like it won't accept requests unless `constants.APIDomain` is specified. I'd rather not dive too deep into that since the HTTP client is being deprecated anyways.

I am trying to find a better way to handle background dialing of the gRPC client though. The `NewAddrsDialer` does not work as intended for gRPC and definitely needs a better solution.

Reply:
I refactored the `connectInBackground` function to always take the first address or the dialer if it's provided, which is how it is being used across teleport. The second address can't be used without adding our own background dialing logic.

I could add something like this to `connectInBackground` to actually use all of the config provided:
```go
// Start a goroutine to test connections and replace the current
// client connection in the background.
go func() {
	var c *Client
	c, err = connect(ctx, cfg)
	if err == nil {
		clt.Close()
		*clt = *c
	} else {
		clt.dialer = ContextDialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
			return nil, err
		})
		clt.dialGRPC(ctx, "", false)
	}
}()
```
A lot would be needed to make this more thread safe, if that's even possible. IIUC dialing in the background is inherently a data race, unless used carefully by the user. However I don't think this is something we need to support outside of how it's already being used.